rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+1. Read the latest manifest version V. (If the writer is applying an update from the local state then in the first iteration it can assume that it’s local version is the latest. If it’s wrong, the CAS will fail and it should reload the manifest from Object Storage))
+2. Decide whether it needs to re-establish it’s checkpoint. The writer reestablishes the checkpoint whenever the set of L0 SSTs, or Sorted Runs has changed. This includes both changes made locally by the Writer, or by the Compactor.
+    1. If so, create a new checkpoint with `manifest_id` set to V+1, `checkpoint_expire_time` set to None, and `metadata` set to `WriterCheckpoint{ epoch: E }`.
+    2. Delete the old checkpoint from `checkpoints`. When we support range scans, we’ll instead want to maintain references to the checkpoint from the range scans and delete the checkpoint when all references are dropped. (We should technically do this for point lookups too, but the likelihood of GC deleting an SST in this case is very low).
+    3. Apply any local changes
+    4. Write the new manifest. If this fails with a CAS error, go back to 1.
+
+#### Clones
+
+The writer will also support initializing a new database from an existing checkpoint. This allows users to “fork” an instance of slatedb, allowing it to access all the original db’s data from the checkpoint, but isolate writes to a new db instance.
+
+To support this, we add the `Db::open_from_checkpoint` method, which accepts a `source_path` and `source_checkpoint`. When initializing the db for the first time (detected by the presence of a manifest), the writer will do the following:
+1. Read the source manifest MC at `source_checkpoint`.
+2. Copy over any WAL SSTs between `last_compacted_wal_id` and `wal_id_last_seen`
+3. Verify that `source_checkpoint` is present and not expired. Create a new checkpoint in the source database with the `manifest_id` as `source_checkpoint`.
+4. Compute an initial manifest for the new DB. The new manifest will set its fields as follows
+    - writer_epoch: set to the epoch from the source checkpoint + 1.
+    - compactor_epoch: copied from the source checkpoint.
+    - wal_id_last_compacted: copied from the source checkpoint
+    - wal_id_last_seen: copied from the source checkpoint
+    - l0: copied from the source checkpoint
+    - l0_last_compacted: copied from the source checkpoint
+    - compacted: copied from the source checkpoint
+    - checkpoints: contains a single entry with the writer’s checkpoint.
+5. If CAS fails, go back to (1) (TODO: we can skip step 2 - we just need to re-validate that the source checkpoint is still valid because the GC could clean it up)
```

Comment:
don't we need some mechanism for ensuring a GC process for the old slate instance doesn't clean up SSTs referenced from the cloned database once the checkpoint it cloned from expires? IIUC, a GC process that is triggered on the original manifest path would only consider manfiests from the original instance.

Reply:
Ah I think I missed this step tucked between all the other comments:

> Create a new checkpoint in the source database with the `manifest_id` as `source_checkpoint`.

I guess the new instance will maintain the expiry of that source checkpoint?

Reply:
> I guess the new instance will maintain the expiry of that source checkpoint?

No, the clone would create a checkpoint with no expiry and only delete that checkpoint as part of destroying the clone.